File: ts/tests/primitives.test.ts
import { describe, it, expect } from "vitest";
import { createAllPrimitives } from "../src/primitives.js";
import type { Primitive, Snapshot } from "../src/primitive.js";
import type { Event } from "../src/event.js";

describe("primitives", () => {
  const all = createAllPrimitives();

  it("createAllPrimitives returns 201 primitives", () => {
    expect(all.length).toBe(201);
  });

  it("all 201 primitives instantiate with valid id", () => {
    const ids = new Set<string>();
    for (const p of all) {
      const id = p.id();
      expect(id.value).toBeTruthy();
      ids.add(id.value);
    }
    // All unique
    expect(ids.size).toBe(201);
  });

  it("Layer 0 has 45 primitives", () => {
    expect(all.filter((p) => p.layer().value === 0).length).toBe(45);
  });

  it("Layer 1 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 1).length).toBe(12);
  });

  it("Layer 2 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 2).length).toBe(12);
  });

  it("Layer 3 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 3).length).toBe(12);
  });

  it("Layer 4 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 4).length).toBe(12);
  });

  it("Layer 5 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 5).length).toBe(12);
  });

  it("Layer 6 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 6).length).toBe(12);
  });

  it("Layer 7 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 7).length).toBe(12);
  });

  it("Layer 8 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 8).length).toBe(12);
  });

  it("Layer 9 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 9).length).toBe(12);
  });

  it("Layer 10 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 10).length).toBe(12);
  });

  it("Layer 11 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 11).length).toBe(12);
  });

  it("Layer 12 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 12).length).toBe(12);
  });

  it("Layer 13 has 12 primitives", () => {
    expect(all.filter((p) => p.layer().value === 13).length).toBe(12);
  });

  it("every primitive has non-empty subscriptions", () => {
    for (const p of all) {
      const subs = p.subscriptions();
      expect(subs.length).toBeGreaterThan(0);
      for (const s of subs) {
        expect(s.value).toBeTruthy();
      }
    }
  });

  it("every primitive has cadence >= 1", () => {
    for (const p of all) {
      expect(p.cadence().value).toBeGreaterThanOrEqual(1);
    }
  });

  it("process returns mutations with correct structure", () => {
    const fakeSnapshot: Snapshot = {
      tick: 1,
      primitives: new Map(),
      pendingEvents: [],
      recentEvents: [],
    };

    for (const p of all) {
      const mutations = p.process(1, [], fakeSnapshot);
      expect(mutations.length).toBe(2);

      const m0 = mutations[0];
      expect(m0.kind).toBe("updateState");
      if (m0.kind === "updateState") {
        expect(m0.key).toBe("eventsProcessed");
        expect(m0.value).toBe(0);
      }

      const m1 = mutations[1];
      expect(m1.kind).toBe("updateState");
      if (m1.kind === "updateState") {
        expect(m1.key).toBe("lastTick");
        expect(m1.value).toBe(1);
      }
    }
  });

  it("process records event count correctly", () => {
    const fakeSnapshot: Snapshot = {
      tick: 5,
      primitives: new Map(),
      pendingEvents: [],
      recentEvents: [],
    };

    // Use a non-empty events array (cast to avoid needing real Event objects)
    const fakeEvents = [{}, {}, {}] as unknown as Event[];
    const p = all[0];
    const mutations = p.process(5, fakeEvents, fakeSnapshot);

    if (mutations[0].kind === "updateState") {
      expect(mutations[0].value).toBe(3);
    }
    if (mutations[1].kind === "updateState") {
      expect(mutations[1].value).toBe(5);
    }
  });
});

File: rust/tests/primitives_test.rs
use std::collections::HashMap;

use eventgraph::primitive::Snapshot;
use eventgraph::primitives::{create_all_primitives, layer_counts};

#[test]
fn test_all_201_primitives_instantiate() {
    let prims = create_all_primitives();
    assert_eq!(prims.len(), 201);
    for p in &prims {
        // Each primitive should have a non-empty id
        assert!(!p.id().value().is_empty(), "Primitive has empty id");
    }
}

#[test]
fn test_create_all_primitives_returns_201() {
    let prims = create_all_primitives();
    assert_eq!(prims.len(), 201, "Expected exactly 201 primitives");
}

#[test]
fn test_layer_counts_correct() {
    let prims = create_all_primitives();
    let mut counts: HashMap<u8, usize> = HashMap::new();
    for p in &prims {
        *counts.entry(p.layer().value()).or_insert(0) += 1;
    }

    for (layer, expected) in layer_counts() {
        let actual = counts.get(&layer).copied().unwrap_or(0);
        assert_eq!(
            actual, expected,
            "Layer {} expected {} primitives, got {}",
            layer, expected, actual
        );
    }

    // Verify all 14 layers are present
    assert_eq!(counts.len(), 14, "Expected 14 layers");
}

#[test]
fn test_process_returns_mutations() {
    let prims = create_all_primitives();
    let snapshot = Snapshot {
        tick: 1,
        primitives: HashMap::new(),
        pending_events: vec![],
        recent_events: vec![],
    };

    for p in &prims {
        let mutations = p.process(1, &[], &snapshot);
        assert_eq!(
            mutations.len(),
            2,
            "Primitive {} should return 2 mutations (eventsProcessed + lastTick)",
            p.id().value()
        );
    }
}

#[test]
fn test_subscriptions_non_empty() {
    let prims = create_all_primitives();
    for p in &prims {
        let subs = p.subscriptions();
        assert!(
            !subs.is_empty(),
            "Primitive {} has empty subscriptions",
            p.id().value()
        );
    }
}

#[test]
fn test_unique_primitive_ids() {
    let prims = create_all_primitives();
    let mut ids: Vec<String> = prims.iter().map(|p| p.id().value().to_string()).collect();
    let total = ids.len();
    ids.sort();
    ids.dedup();
    assert_eq!(ids.len(), total, "All primitive IDs must be unique");
}

#[test]
fn test_cadence_is_valid() {
    let prims = create_all_primitives();
    for p in &prims {
        assert!(
            p.cadence().value() >= 1,
            "Primitive {} has invalid cadence",
            p.id().value()
        );
    }
}

#[test]
fn test_layer_values_in_range() {
    let prims = create_all_primitives();
    for p in &prims {
        assert!(
            p.layer().value() <= 13,
            "Primitive {} has layer > 13",
            p.id().value()
        );
    }
}
